src/adapters/vertexai/llm.ts

```diff
@@ -0,0 +1,159 @@
+  BaseLLMTokenizeOutput,
+  ExecutionOptions,
+  GenerateCallbacks,
+  GenerateOptions,
+  LLMCache,
+  LLMMeta,
+} from "@/llms/base.js";
+import { shallowCopy } from "@/serializer/utils.js";
+import type { GetRunContext } from "@/context.js";
+import { Emitter } from "@/emitter/emitter.js";
+import { VertexAI } from "@google-cloud/vertexai";
+import { Role } from "@/llms/primitives/message.js";
+import { signalRace } from "@/internals/helpers/promise.js";
+import { processContentResponse, registerVertexAI, createModel } from "./utils.js";
+
+interface VertexAILLMChunk {
+  text: string;
+  metadata: Record<string, any>;
+}
+
+export class VertexAILLMOutput extends BaseLLMOutput {
+  public readonly chunks: VertexAILLMChunk[] = [];
+
+  constructor(chunk: VertexAILLMChunk) {
+    super();
+    this.chunks.push(chunk);
+  }
+
+  merge(other: VertexAILLMOutput): void {
+    this.chunks.push(...other.chunks);
+  }
+
+  getTextContent(): string {
+    return this.chunks.map((result) => result.text).join("");
+  }
+
+  toString(): string {
+    return this.getTextContent();
+  }
+
+  createSnapshot() {
+    return { chunks: shallowCopy(this.chunks) };
+  }
+
+  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>): void {
+    Object.assign(this, snapshot);
+  }
+}
+
+export interface VertexAILLMInput {
+  modelId: string;
+  project: string;
+  location: string;
+  client?: VertexAI;
+  executionOptions?: ExecutionOptions;
+  cache?: LLMCache<VertexAILLMOutput>;
+  parameters?: Record<string, any>;
+}
+
+export class VertexAILLM extends LLM<VertexAILLMOutput, GenerateOptions> {
+  public readonly emitter: Emitter<GenerateCallbacks> = Emitter.root.child({
+    namespace: ["vertexai", "llm"],
+    creator: this,
+  });
+
+  protected client: VertexAI;
+
+  constructor(protected readonly input: VertexAILLMInput) {
+    super(input.modelId, input.executionOptions, input.cache);
+    this.client =
+      input.client ?? new VertexAI({ project: input.project, location: input.location });
+  }
+
+  static {
+    this.register();
+    registerVertexAI();
+  }
+
+  async meta(): Promise<LLMMeta> {
+    return { tokenLimit: Infinity };
+  }
+
+  async tokenize(input: LLMInput): Promise<BaseLLMTokenizeOutput> {
+    const generativeModel = createModel(this.client, this.modelId);
+    const response = await generativeModel.countTokens({
+      contents: [{ parts: [{ text: input }], role: Role.USER }],
+    });
+    return {
+      tokensCount: response.totalTokens,
+    };
+  }
+
+  protected async _generate(
+    input: LLMInput,
+    options: GenerateOptions,
+    run: GetRunContext<this>,
+  ): Promise<VertexAILLMOutput> {
+    const generativeModel = createModel(this.client, this.modelId, options.guided?.json);
+    const responses = await signalRace(() => generativeModel.generateContent(input), run.signal);
+    const result: VertexAILLMChunk = {
+      text: processContentResponse(responses.response),
+      metadata: { createdAt: new Date() },
```

Comment:
I thought metadata from the `responses` variable like `usageMetadata`. 

Note that those values should have somehow defined "merging" behavior in your `VertexAIChatLLMOutput` (this applies only to streaming, I would say - for instance, you need to sum token spends).